### ISSUE TITLE
Review pane contains edits that took place outside of the current merge request

### ISSUE THREAD

**Issue by dav1d-wright** [harrisoncramer/gitlab.nvim#63]
## Prerequsities

- [x] The "Troubleshooting" section of the README did not help
- [x] I've installed the required dependencies
- [x] I'm on the latest version of the plugin

## Bug Description
When looking at the diff in gitlab.nvim I see files that are marked as modified which have been edited in the meantime and merged into the target branch, but not as part of the branch that is being reviewed. The current branch has not yet been rebased onto the target branch, which is probably part of why this is marked as a diff.

## Reproduction Steps

I believe the following should reproduce this issue:

1. Create a branch off of an older version of the target branch 
2. Create some modifications and commit them
3. Create a pull request
4. Open gitlab.cn

## Screenshots
Nvim:

![image](https://github.com/harrisoncramer/gitlab.nvim/assets/3441341/869f24ff-4023-4af7-8d06-bb5fa2649113)

GitLab:

![image](https://github.com/harrisoncramer/gitlab.nvim/assets/3441341/05a5a02e-82c9-4988-b61d-69fe98f90585)

## Other Details
It seems that gitlab.nvim displays the diff between the target and the source branch if I'm not mistaken, because after I checked out the target branch and pulled from `origin`, the diff in the review pane looks different. However, this may contain changes that were not part of the current pull request. This differs from what GitLab displays. Maybe it would be possible to fetch the Diff from GitLab itself? 

## Finally
Thank you very much for creating this tool! I greatly appreciate it. It looks like exactly what I've been hoping for but didn't have the time or expertise to write myself. If there is anything that I can help with in reproducing or fixing this then just let me know.

**Comment by harrisoncramer** [comment]
Thanks for reporting this, and for the kind words.

The reviewers for this plugin (Diffview and Delta) both operate on the local filesystem, so we'd have to find a way to pull down and even merge the target branch into the feature branch prior to doing the comparison.

I've got an MR open here to do the first step of that: To detect when your local branch is out-of-date compared to the branch that's remote, and will prompt you to pull down those changes.

https://github.com/harrisoncramer/gitlab.nvim/pull/61

However, you would still need to merge the target branch into your feature branch in order to see the difference between the two. Realistically I think that the simplest path here would be to (after that above MR is merged) add the additional step of merging the target branch into your feature branch prior to opening up the reviewer. 

It looks like the online UI uses an external service called Gitaly and Redis to do some comparisons and parsing which isn't really feasible locally:

https://docs.gitlab.com/ee/development/merge_request_concepts/diffs/

**Comment by droopy4096** [comment]
actually based on behaviour from VSCode and GitLab itself - diff needs to be generated from the point of forking to the tip of the branch and not source_branch->target_branch.

**Comment by briandipalma** [comment]
If I could make gitlab.nvim connect to the view I launch with the command up above that would be perfect for me for 99% of reviews.

**Comment by harrisoncramer** [comment]
Thank you both. I've been doing a bit of research (thanks ChatGPT) and it seems like the following series of commands may provide us with the exact commit hashes we need to provide to the `diffview` plugin in order to generate the correct output. Please let me know if this sounds right to you, and I can work on implementing this.

1) Pull down the latest information from the remote repository:

```
$ git fetch origin
```

2) Use the `git merge-base` command to get the point two branches diverged. This seems like a simpler alternative than the approach that @droopy4096 outlined:

```
git merge-base origin/target-branch origin/source-branch
```

3) Get the _last commit in the feature branch_ that's been pushed up to the origin remote:

```
$ git rev-parse origin/feature-branch
```

4) Run the `diffview` command with these two hashes:

```
:DiffviewOpen common_ancestor...last_commit
```

5) If there are merge conflicts with the branch (possible) then show a warning to the user. We already have merge conflict warning information in the project already, but I'm thinking it makes sense to flash a warning to the user if they open the reviewer and there are conflicts in the branch.

**Comment by harrisoncramer** [comment]
Ohhh, wait on second thought it looks like these commit hashes might already be available to us via the information response:

https://docs.gitlab.com/ee/api/merge_requests.html#response

They look like they are in the `diff_refs` field here? It seems like we'd just have to compare these fields:

```
:DiffviewOpen (diff_refs.start_sha)..(diff_refs.head_sha)
```

Would this work @droopy4096 @briandipalma?

**Comment by briandipalma** [comment]
Your change doesn't seem to be in main any longer?

https://github.com/harrisoncramer/gitlab.nvim/blob/main/lua/gitlab/reviewer/diffview.lua

This is weird...you're change didn't work so I thought I'd open up the file and make some changes to the code you changed to see if I could fix it and I can't see your change in there?!

I must be making a silly mistake, is that the wrong file?

**Comment by harrisoncramer** [comment]
Oh, the change wass still in a PR, I hadn't merged it yet. It's now in main, if you update the plugin you'll see it now.

**Comment by briandipalma** [comment]
Fantastic, it works as I'd expect, thanks!

**Comment by harrisoncramer** [comment]
Great! I'm going to close this issue out but anyone noticing similar issues like this please open up another one. Thanks all for your help figuring this one out!
